core/drivers/sp805_wdt.c

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019, Broadcom
+ */
+
+#include <drivers/sp805_wdt.h>
+#include <initcall.h>
+#include <io.h>
+#include <keep.h>
+#include <kernel/interrupt.h>
+#include <mm/core_memprot.h>
+#include <stdint.h>
+#include <trace.h>
+
+static vaddr_t chip_to_base(struct wdt_chip *chip)
+{
+	struct sp805_wdt_data *pd =
+		container_of(chip, struct sp805_wdt_data, chip);
+
+	return io_pa_or_va(&pd->base);
+}
+
+static TEE_Result sp805_setload(struct wdt_chip *chip, unsigned long timeout)
+{
+	struct sp805_wdt_data *pd =
+		container_of(chip, struct sp805_wdt_data, chip);
+	uint32_t load = 0;
+
+	/*
+	 * sp805 runs counter with given value twice, after the end of first
+	 * counter it gives an interrupt and then starts counter again. If
+	 * interrupt already occurred then it resets the system. This is why
+	 * load is half of what should be required.
+	 */
+	if (MUL_OVERFLOW(timeout, pd->clk_rate, &load))
+		return TEE_ERROR_SECURITY;
+
+	load =  (load / 2) - 1;
+	if (load < WDT_LOAD_MIN)
+		load = WDT_LOAD_MIN;
+
+	pd->load_val = load;
+	return TEE_SUCCESS;
+}
+
+static void sp805_config(struct wdt_chip *chip, bool ping)
```

Comment:
minor: maybe rename `ping` into `enable` as this seems what the argument is used for.

Reply:
Done.